client/widgets/CArtifactHolder.cpp

```diff
@@ -311,6 +311,9 @@ void CHeroArtPlace::select ()
 void CHeroArtPlace::deselect ()
 {
 	pickSlot(false);
+	selectSlot(false);
+	CBonusSystemNode::treeHasChanged();
```

Comment:
Do not do that, this is a dirty workaround, you should call `treeHasChanged` only if bonus tree was actually changed and bonus tree should be changed only by applying NetPacks.

Reply:
I think this is the most straight forward change. If not do this, we might need to clean up the cachingStr data in the cache, but there is no function to do that, currently only changing the tree would clean the cache. 
Since the "deselect" is not a very heavily called action, it would be fine to put a special case here. I think this is the simplest way to fix it.
Could you propose any other approaches? I think "deselect" is a special case that the coder didn't consider for the tree change, there will not be other cases like this.

Reply:
The most straightforward fix is to ignore cache in `CHeroWithMaybePickedArtifact::getAllBonuses` just pass empty string to `hero`. 
The most simple but correct fix is to modify `CHeroWithMaybePickedArtifact::getAllBonuses` to not change `heroBonuses` and change `out` instead.

Reply:
I used your first approach, and validated it's working fine, so I pushed that change.
For your second approach, I think the problem is inside of the **heroBonuses**, so **getAllBonuses** cannot get correct **out** if **heroBonuses** is already wrong.

Reply:
The actual bug is that we modify `heroBonuses`, but should modify `out` instead.
```
for(auto b : *bonusesFromPickedUpArtifact)
		*heroBonuses -= b;
```

Reply:
Oh no, this is not the root cause. I know that the heroBonuses should be const, and we should modify the "out". I could update this part, but that would not help resolve this issue.
During my debugging, the real issue is, the "heroBonuses = hero->getAllBonuses(selector, limit, hero);" is already returning the wrong result (because it's reading cached value after it's picked up).

Reply:
I have made the code swap you mentioned, that we should update the "out" not "heroBonuses".

Reply:
I suppose what you want to say is, to update the current logic:
   heroBonuses = getAllBonuses(...)
   for(auto b : *bonusesFromPickedUpArtifact)
	 out -= b

to new logic:
   heroBonuses = getAllBonuses(...)
   if (pickedUp)
      for(auto b : *bonusesFromPickedUpArtifact)
	 out -= b
   else  // Put back
    for(auto b : *bonusesFromPickedUpArtifact)
	 out += b

That would possibly resolve the issue, but I just tried to update like that, and found that in the deselect() function, the pickedUpArtifact is already called "clear()" before call getAllBonuses, so this proposal might still have problem.

Reply:
> I have made the code swap you mentioned, that we should update the "out" not "heroBonuses".

And this should be enough

Reply:
Thanks. so could you help me complete this PR?